src/ast_editor.ml

```diff
@@ -0,0 +1,585 @@
+    if Ojs.has_property msg name then
+      Some (Int.of_string (Ojs.string_of_js (Ojs.get_prop_ascii msg name)))
+    else
+      None
+  in
+  match int_prop "selectedOutput" with
+  | Some i ->
+    let webview_opt =
+      let ast_editor_state = Extension_instance.ast_editor_state instance in
+      Ast_editor_state.find_webview_by_doc ast_editor_state
+        (TextDocument.uri document)
+    in
+    Ast_editor_state.set_original_mode ast_editor_state (i = 0);
+    refresh_ast_explorer instance ~document ~webview_opt
+  | None -> (
+    match Option.both (int_prop "begin") (int_prop "end") with
+    | None -> ()
+    | Some (cbegin, cend) ->
+      let apply_selection editor cbegin cend =
+        let document = TextEditor.document editor in
+        let anchor = Vscode.TextDocument.positionAt document ~offset:cbegin in
+        let active = Vscode.TextDocument.positionAt document ~offset:cend in
+        TextEditor.set_selection editor
+          (Selection.makePositions ~anchor ~active);
+        TextEditor.revealRange editor
+          ~range:(Range.makePositions ~start:anchor ~end_:active)
+          ()
+      in
+      Vscode.Window.visibleTextEditors ()
+      |> List.iter ~f:(fun editor ->
+             let visible_doc = TextEditor.document editor in
+             if (* !original_mode && *) document_eq document visible_doc then
+               apply_selection editor cbegin cend
+             else if
+               (* (not !original_mode) && *)
+               (Ast_editor_state.entry_exists ast_editor_state
+                  ~origin_doc:(TextDocument.uri document))
+                 ~pp_doc:(TextDocument.uri visible_doc)
+               && not (Ast_editor_state.get_original_mode ast_editor_state)
+             then
+               match Option.both (int_prop "r_begin") (int_prop "r_end") with
+               | None -> ()
+               | Some (rcbegin, rcend) -> apply_selection editor rcbegin rcend))
+
+let on_hover custom_doc webview =
+  let provider =
+    let provideHover ~(document : TextDocument.t) ~(position : Position.t)
+        ~token:_ =
+      let offset = TextDocument.offsetAt document ~position in
+      if document_eq custom_doc document then
+        send_msg "focus" (Ojs.int_to_js offset) ~webview;
+      let hover =
+        Hover.make
+          ~contents:(`MarkdownString (MarkdownString.make ~value:"" ()))
+      in
+      `Value (Some hover)
+    in
+    HoverProvider.create ~provideHover
+  in
+  Vscode.Languages.registerHoverProvider ~selector:(`String "ocaml") ~provider
+
+let activate_hover_mode instance ~document =
+  let ast_editor_state = Extension_instance.ast_editor_state instance in
+  match
+    Ast_editor_state.find_webview_by_doc ast_editor_state
+      (TextDocument.uri document)
+  with
+  | Some webview -> on_hover document webview
+  | None ->
+    show_message `Error "Webview wasn't found while switching hover mode";
+    failwith "Webview wasn't found while switching hover mode"
```

Comment:
This is a fatal error that the user isn't responsible for and we cannot recover from, right? In that case we should introduce an exception type called `Code_error`. In general, we shouldn't be using `failwith` at all.

Reply:
This is not a fatal error, it just means that user tried to activate hover mode without opening an AST editor. The `failwith` is useless though 8e2c3055024505f21f2bea2b707ef23ee43cf61b